packages/lodestar/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -265,64 +271,80 @@ export class MatchingDataAttestationGroup {
+          // this new attestation is actually a subset of an existing one, don't want to add it
+          return InsertOutcome.AlreadyKnown;
+
+        case IntersectResult.Exclusive:
+          // no intersection
+          aggregateInto(prevAttestation, attestation);
+          return InsertOutcome.Aggregated;
+
+        case IntersectResult.Superset:
+          // newBits superset of prevBits
+          // this new attestation is superset of an existing one, remove existing one
+          indicesToRemove.push(i);
       }
     }
-    if (insertResult === InsertOutcome.NewData) {
-      for (const index of indicesToRemove.reverse()) {
-        this.attestations.splice(index, 1);
-      }
-      this.attestations.push(attestation);
-      // Remove the attestations with less participation
-      if (this.attestations.length > MAX_RETAINED_ATTESTATIONS_PER_GROUP) {
-        this.attestations.sort((a, b) => b.attestingIndices.size - a.attestingIndices.size);
-        this.attestations.splice(
-          MAX_RETAINED_ATTESTATIONS_PER_GROUP,
-          this.attestations.length - MAX_RETAINED_ATTESTATIONS_PER_GROUP
-        );
-      }
+
+    // Added new data
+    for (const index of indicesToRemove.reverse()) {
+      // TODO: .splice performance warning
+      this.attestations.splice(index, 1);
     }
-    return insertResult;
+
+    this.attestations.push(attestation);
+
+    // Remove the attestations with less participation
+    if (this.attestations.length > MAX_RETAINED_ATTESTATIONS_PER_GROUP) {
+      this.attestations.sort((a, b) => b.trueBitsCount - a.trueBitsCount);
+      this.attestations.splice(
+        MAX_RETAINED_ATTESTATIONS_PER_GROUP,
+        this.attestations.length - MAX_RETAINED_ATTESTATIONS_PER_GROUP
+      );
+    }
+
+    return InsertOutcome.NewData;
   }
 
-  getAttestationsForBlock(seenAttestingIndices: Set<ValidatorIndex>): AttestationWithIndex[] {
-    const attestations: AttestationWithIndex[] = [];
+  getAttestationsForBlock(seenAttestingIndices: Set<ValidatorIndex>): AttestationNonParticipant[] {
+    const attestations: AttestationNonParticipant[] = [];
+
+    const committeeLen = this.committee.length;
+    const committeeSeenAttesting = new Array<boolean>(committeeLen);
 
-    for (const attestation of this.attestations) {
+    // Intersect committee with participation only once for all attestations
+    for (let i = 0; i < committeeLen; i++) {
+      committeeSeenAttesting[i] = seenAttestingIndices.has(this.committee[i]);
+    }
+
+    for (const {attestation} of this.attestations) {
+      const {aggregationBits} = attestation;
       let notSeenAttesterCount = 0;
-      for (const attIndex of attestation.attestingIndices) {
-        if (!seenAttestingIndices.has(attIndex)) notSeenAttesterCount++;
+
+      for (let i = 0; i < committeeLen; i++) {
+        // TODO: Optimize aggregationBits.get() in bulk for the entire BitArray
+        if (!committeeSeenAttesting[i] && aggregationBits.get(i)) {
```

Comment:
this is the hot path. If we store `attestingIndices: number[]` in `AttestationWithIndex`, and change `seenAttestingIndices` param to `number[]` and make sure they are in the order of the committee,  this could be faster?
```typescript
let seenCount = 0;
let checkIndex = 0;
for (let i = 0; i < seenAttestingIndices.length; i++) {
  if (checkIndex >= attestingIndices.length) break;
  if(seenAttestingIndices[i] === attestingIndices[checkIndex) {
    seenCount ++;
    checkIndex++;
  }
}
const notSeenAttesterCount = attestingIndices.length - seenCount;
```

Reply:
nvm, this is incorrect